### PR TITLE
TN-4316: Export versioned packages.

### PR DESCRIPTION
**Pull request by fisgeci**
#### Description

When exporting packages and their dependencies, currently we only export the latest version of each package instead of the required version that some package along the chain depends on. 
This pr makes it so we export the used version of each package and also sets the field in the usedVersion field in the manifest for each package.

#### Checklist

- [x] I have self-reviewed this PR
- [x] I have tested the change and proved that it works in different scenarios
- [x] I have updated docs if needed

### CODE REVIEW

File: src/api/package-api.ts [hunk 1]
```diff
@@ -15,8 +15,8 @@ class PackageApi {
         });
     }
 
-    public async exportPackage(rootPackageKey: string): Promise<Buffer> {
-        return await httpClientV2.downloadFile(`/package-manager/api/packages/${rootPackageKey}/export?newKey=${rootPackageKey}`);
+    public async exportPackage(rootPackageKey: string, version?: string): Promise<Buffer> {
+        return await httpClientV2.downloadFile(`/package-manager/api/packages/${rootPackageKey}/export?newKey=${rootPackageKey}${version ? `&version=${version}` : ""}`);
```
Comment (AlbinHoxha) [inline src/api/package-api.ts:19]: can you look to use `querystring.stringify`to build parameters like we did in some other methods

File: src/commands/package.command.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ export class PackageCommand {
         await packageService.batchExportPackages(packageKeys, includeDependencies);
     }
 
-    public async batchImportPackages(spaceMappings: string[], exportedDatapoolsFile: string, exportedPackagesFile: string): Promise<void> {
-        await packageService.batchImportPackages(spaceMappings ?? [], exportedDatapoolsFile, exportedPackagesFile);
+    public async batchImportPackages(spaceMappings: string[], dataModelMapping: string, exportedPackagesFile: string): Promise<void> {
```
Comment (AlbinHoxha) [inline src/commands/package.command.ts:51]: I think this should be a list of strings

File: src/content-cli-import.ts [hunk 3]
```diff
@@ -12,10 +12,10 @@ export class Import {
             .description("Command to import all given packages")
             .option("-p, --profile <profile>", "Profile which you want to use to list packages")
             .option("--spaceMappings <spaceMappings...>", "List of mappings for importing packages to different target spaces. Mappings should follow format 'packageKey:targetSpaceKey'")
-            .option("--exportedDatapoolsFile <exportedDatapoolsFile>", "Exported datapool file (relative path)", "")
+            .option("--dataModelMapping <dataModelMapping>", "Data model mappings file (relative path)", "")
```
Comment (AlbinHoxha) [inline src/content-cli-import.ts:15]: did we not decide we will be sending a mapping like the spaceMapping and not the file?

File: src/services/package-manager/package-service.ts [hunk 4]
```diff
@@ -264,40 +263,43 @@ class PackageService {
...
+
+        const fileName = "export_" + uuidv4() + ".zip";
+        zip.writeZip(fileName);
+        logger.info("Successfully exported package to: " + fileName);
     }
 
-    private exportManifestOfPackages(nodes: BatchExportNodeTransport[]): ManifestNodeTransport[] {
-        const manifestNodes: ManifestNodeTransport[] = [];
+    private exportManifestOfPackages(nodes: BatchExportNodeTransport[], dependencyVersionsByNodeKey: Map<string, string[]>): ManifestNodeTransport[] {
+        const manifestNodesByPackageKey = new Map<string, ManifestNodeTransport>();
```
Comment (AlbinHoxha) [inline src/services/package-manager/package-service.ts:297]: why is it better to have this as a map compared to a list?

File: src/services/package-manager/package-service.ts [hunk 5]
```diff
@@ -264,40 +263,43 @@ class PackageService {
...
     }
 
-    private async exportPackagesAndAssets(nodes: BatchExportNodeTransport[]): Promise<any[]> {
+    private async exportPackagesAndAssets(nodes: BatchExportNodeTransport[], versionsByNodeKey: Map<string, string[]>): Promise<any[]> {
         const zips = [];
-        const packages = nodes as ContentNodeTransport[];
-        for (const rootPackage of packages) {
-            const exportedPackage = await packageApi.exportPackage(rootPackage.key)
+        for (const rootPackage of nodes) {
+            let exportedPackage;
```
Comment (AlbinHoxha) [inline src/services/package-manager/package-service.ts:269]: why do we need to declare it here instead of one line below

File: src/services/package-manager/package-service.ts [hunk 6]
```diff
@@ -141,121 +140,121 @@ class PackageService {
...
+                    dependencyVersions.push(dependency.version);
+                    versionsByNodeKey.set(dependency.key, dependencyVersions);
+                }
+            });
 
-        nodesListToExport = await this.getNodeDependencies(nodesListToExport, allPackages, actionFlowPackageKeys, nodePath);
-        return nodesListToExport
-    }
+            const nodesToGetKeys = node.dependencies.filter(dependency => {
+                return !this.nodeHasBeenResolvedBefore(resolvedDependencies, dependency.key, dependency.version)
```
Comment (AlbinHoxha) [inline src/services/package-manager/package-service.ts:233]: Did we decide that we are safe from circular dependency and do not need to add a check that kills the process

File: src/services/package-manager/package-service.ts [hunk 7]
```diff
@@ -141,121 +140,121 @@ class PackageService {
...
-        importedKeys.push(packageToImport.packageKey);
-        const nodeInTargetTeam = await nodeApi.findOneByKeyAndRootNodeKey(packageToImport.packageKey, packageToImport.packageKey);
-        await packageApi.importPackage(packageZip, targetSpaceId, !!nodeInTargetTeam);
-        if (nodeInTargetTeam) {
-            await packageApi.movePackageToSpace(nodeInTargetTeam.id, targetSpaceId)
-        }
-        await this.updateDependencyVersions(packageToImport);
-        await this.publishPackage(packageToImport);
-        logger.info(`Imported package with key: ${packageToImport.packageKey} successfully`)
+      // TODO [TN-4317](https://celonis.atlassian.net/browse/TN-4317)
```
Comment (AlbinHoxha) [inline src/services/package-manager/package-service.ts:143]: This pr is not related to import hence we do not need to comment everything out

File: src/content-cli-import.ts (hunk 3 above)
Comment (fisgeci) [reply src/content-cli-import.ts:15]: We will recieve the mapping in the file. As opposed to proving the exported datapools as before.

File: src/services/package-manager/package-service.ts (hunk 4 above)
Comment (fisgeci) [reply src/services/package-manager/package-service.ts:297]: It is faster to search by key when we need to update the version of each package.

File: src/services/package-manager/package-service.ts (hunk 7 above)
Comment (fisgeci) [reply src/services/package-manager/package-service.ts:143]: I can just delete the import implementation as this pr breaks the current implementation. Ill just leave the comment that links the ticket.

File: src/services/package-manager/package-service.ts (hunk 5 above)
Comment (fisgeci) [reply src/services/package-manager/package-service.ts:269]: This was left over.

File: src/services/package-manager/package-service.ts [hunk 8]
```diff
@@ -266,38 +203,40 @@ class PackageService {
...
         for (const packageZip of packageZips) {
-            zip.addFile(packageZip.packageKey + ".zip", packageZip.data)
+            zip.addFile(`${packageZip.packageKey}_${packageZip.version}.zip`, packageZip.data)
         }
-        zip.writeZip("export_" + uuidv4() + ".zip");
-        logger.info("Successfully exported package");
+
+        const fileName = "export_" + uuidv4() + ".zip";
+        zip.writeZip(fileName);
+        logger.info("Successfully exported package to: " + fileName);
```
Comment (frkntrcnn) [inline src/services/package-manager/package-service.ts:230]: To stick with the structure, how would it be to use "File downloaded successfully. New filename: " this message on output?

File: src/interfaces/manifest-transport.ts [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
 
 export interface ManifestNodeTransport {
+    packageKeyAndVersion: string;
     packageKey: string;
     packageId: string,
-    packageVersion: string;
     variables: ManifestVariable[],
     space: ManifestSpace,
-    dependencies: ManifestDependency[]
+    dependenciesByVersion: Map<string, ManifestDependency[]>;
```
Comment (frkntrcnn) [inline src/interfaces/manifest-transport.ts:9]: is there any possibility to pass dependency without specifying the version? Or should it be pointed out even if the latest version will be used?

File: src/commands/package.command.ts (hunk 2 above)
Comment (promeris) [reply src/commands/package.command.ts:51]: There should probably be two options, either providing the output file from the import of dataModels (which can be implemented when that is ready), or you can define the mappings yourself like `--variableMappings dm-variable,dm-1;dm-variable2,dm-2`, etc.

File: DOCUMENTATION.md [hunk 10]
```diff
@@ -255,6 +255,22 @@ When you use overwrite the following is to be taken into consideration:
 content-cli push package -p my-profile-name --spaceKey my-space -f <path-to-my-local-package> --overwrite
 ```
 
+### Export multiple packages at once from Studio
+
+You can use the export packages command to export multiple packages at once from studio.
+
+```
+//Exporting multiple packages at once
+content-cli export packages -p <your-chosen-profile> --packageKeys <packageKeysYouWantToExport>
```
Comment (promeris) [inline DOCUMENTATION.md:264]: Something like this maybe?
```suggestion
content-cli export packages -p <your-chosen-profile> --packageKeys <package1> <package2>
```

File: DOCUMENTATION.md [hunk 11]
```diff
@@ -255,6 +255,22 @@ When you use overwrite the following is to be taken into consideration:
...
+```
+
+Example usage would be:
+```
+content-cli export packages -p dev --packageKeys package-1 package-2
+```
+
+You can use the --includeDependencies flag to also export the dependencies of the specified packages.
```
Comment (promeris) [inline DOCUMENTATION.md:272]: Let's add an example command for this as well.

**Review by promeris — APPROVED** [review]
